### PR TITLE
Add ability to auth by email address in addition to username

### PR DESCRIPTION
**Pull request by doubt72**

### CODE REVIEW

**Comment by smith** [comment]
Should we have a `context` in the specs for this?

**Comment by smith** [comment]
:+1:, though a test would be great.

**Comment by doubt72** [comment]
Yeah, think so, but TBH only reason I didn't add one is that there's no testing of actual auth path at all, really, and I wasn't sure how much effort that was worth if it's basically tested via oc-id anyway?

**Comment by smith** [comment]
Fair enough. MWR.

File: lib/omniauth/strategies/chef.rb [hunk 1]
```diff
@@ -59,7 +59,17 @@ def callback_phase
...
+          # Check if username is email
+          if (uname.include?('@'))
+            users = chef.get_rest(
+              "users?#{{ email: username }.to_query}"
+            )
+            if (users.length > 0)
+              uname = users.first[0]
+            end
+          end
+          chef.post_rest(resource, username: uname, password: password)['user']
```
Comment (mmzyk) [inline lib/omniauth/strategies/chef.rb]: What happens if users is 0 length? I guess this problem exists in the old code as well.

Comment (mmzyk) [inline lib/omniauth/strategies/chef.rb]: Oh, looks like it falls to the rescue, which is empty.

Comment (doubt72) [inline lib/omniauth/strategies/chef.rb]: If zero length, it means no email address matches, so it tries as a regular authentication.  It does get caught and will show an invalid credentials error, so it is handled (I think it ends up calling retry if it fails in the sessions controller in OC-ID).

Comment (mmzyk) [inline lib/omniauth/strategies/chef.rb]: Ah, okay, thanks for the explanation.

**Comment by mmzyk** [comment]
:+1: Merge away.

**Pull request merged**